### PR TITLE
ERC721 Approve view

### PR DESCRIPTION
**Pull request by blackdevelopa**
**Development & PR Process**
1. Follow MetaMask [Mobile Coding Standards](https://docs.google.com/document/d/1VJLwTRsUw_5EDq_o8d6sSbXUAYENLurkRitYO45eY5o/edit?usp=sharing)
2. Add `release-xx` label to identify the PR slated for a upcoming release (will be used in release discussion)
3. Add `needs-dev-review` label when work is completed
4. Add `needs-qa` label when dev review is completed
5. Add `QA Passed` label when QA has signed off

**Description**
This PR brings the ERC721 token approve to mobile. Prior to this, the ERC20 approve modal was shown for ERC721 token approve. See [issue](https://github.com/MetaMask/metamask-mobile/issues/5520) for details on the extension and now mobile.

Test Suite:
* [x] ERC20 Approve
* [x] ERC721 Approve
* [x] ERC1155 Approve?

**Screenshots/Recordings**
http://recordit.co/uZIxhi2GUH (Updated)

**Issue**

Progresses #5520 

**Checklist**

* [x] There is a related GitHub issue
* [ ] Tests are included if applicable
* [ ] Any added code is fully documented

### CODE REVIEW

**Comment by bschorchit** [comment]
The NFT screen is looking great, thanks Sylva!
Question: Is the long loading state (see screenshot) because it's a dev build running on an emulator? Or is this expected in prod as well? In the video it seemed it took 9s to update it to the NFT information.

Screenshot 2023-01-24 at 10 36 35

File: app/components/UI/ApproveTransactionReview/index.js [hunk 1]
```diff
@@ -287,14 +300,34 @@ class ApproveTransactionReview extends PureComponent {
...
-          to,
+        const isContract = await isSmartContractAddress(
+          spenderAddress,
+          chainId,
         );
-        tokenSymbol = await AssetsContractController.getERC721AssetSymbol(to);
+        if (!isContract) {
+          const tokenDetails = await getTokenDetails(to, from, encodedValue);
+          if (tokenDetails?.standard === ERC721) {
+            tokenName = tokenDetails.name;
```
Comment (jpuri) [inline app/components/UI/ApproveTransactionReview/index.js:317]: May be `if (tokenDetails?.standard === ERC721 || tokenDetails?.standard === ERC1155)`.

Comment (blackdevelopa) [reply app/components/UI/ApproveTransactionReview/index.js:317]: Done here https://github.com/MetaMask/metamask-mobile/pull/5587/commits/f5f48d41b4dad6f287a1f30932427c341e9c25c1

File: app/components/UI/ApproveTransactionReview/index.js [hunk 2]
```diff
@@ -639,15 +684,39 @@ class ApproveTransactionReview extends PureComponent {
...
+                  (#{token?.tokenId})
+                </Text>
+              </ButtonLink>
+            ) : (
+              <Text variant={TextVariants.sHeadingMD}>{tokenSymbol}</Text>
             )}
           </Text>
-          {originalApproveAmount && (
+
+          {tokenType !== (ERC721 || ERC1155) && originalApproveAmount && (
```
Comment (jpuri) [inline app/components/UI/ApproveTransactionReview/index.js:719]: This should be `tokenType !== ERC721 && tokenType !== ERC1155 ...`

File: app/components/UI/ApproveTransactionReview/index.js [hunk 3]
```diff
@@ -662,14 +731,17 @@ class ApproveTransactionReview extends PureComponent {
...
-          <TouchableOpacity
-            style={styles.actionTouchable}
-            onPress={this.toggleEditPermission}
-          >
-            <Text reset style={styles.editPermissionText}>
-              {strings('spend_limit_edition.edit_permission')}
-            </Text>
-          </TouchableOpacity>
+
+          {fetchingUpdateDone && tokenType !== (ERC721 || ERC1155) && (
```
Comment (jpuri) [inline app/components/UI/ApproveTransactionReview/index.js:735]: Same as above.

File: app/components/UI/ApproveTransactionReview/index.js (hunk 2 above)
Comment (blackdevelopa) [reply app/components/UI/ApproveTransactionReview/index.js]: https://github.com/MetaMask/metamask-mobile/pull/5587/commits/aab9a94c5d80ffc15b479b9604ab4a98067ae338

File: app/components/UI/ApproveTransactionReview/index.js [hunk 4]
```diff
@@ -66,12 +68,20 @@ import formatNumber from '../../../util/formatNumber';
 import { allowedToBuy } from '../FiatOnRampAggregator';
 import { MM_SDK_REMOTE_ORIGIN } from '../../../core/SDKConnect';
 import createStyles from './styles';
+import ButtonLink from '../../../component-library/components/Buttons/Button/variants/ButtonLink';
+import Text, {
+  TextVariants,
+} from '../../../component-library/components/Texts/Text';
```
Comment (jpuri) [inline app/components/UI/ApproveTransactionReview/index.js:74]: Nit: Is order of imports maintained in mobile codebase ?

File: app/components/UI/ApproveTransactionReview/index.js [hunk 5]
```diff
@@ -639,37 +684,68 @@ class ApproveTransactionReview extends PureComponent {
...
                 originIsDeeplink ? 'allow_to_address_access' : 'allow_to_access'
               }`,
-              { tokenSymbol },
+            )}{' '}
+            {!fetchingUpdateDone && (
+              <Text variant={TextVariants.lHeadingMD}>
+                {strings('spend_limit_edition.token')}
+              </Text>
+            )}
+            {tokenType === (ERC721 || ERC1155) ? (
```
Comment (jpuri) [inline app/components/UI/ApproveTransactionReview/index.js:702]: `tokenType === (ERC721 || ERC1155)` I see this pattern at few places.

File: app/util/address/index.js [hunk 6]
```diff
@@ -437,3 +441,25 @@ export const stripHexPrefix = (str) => {
...
+
+export const getTokenDetails = async (tokenAddress, userAddress, tokenId) => {
+  const { AssetsContractController } = Engine.context;
+  const tokenData = await AssetsContractController.getTokenStandardAndDetails(
+    tokenAddress,
+    userAddress,
+    tokenId,
+  );
+  const { standard, name, symbol, decimals } = tokenData;
+  if (standard === (ERC721 || ERC1155)) {
```
Comment (jpuri) [inline app/util/address/index.js:453]: ☝️

File: app/components/UI/ApproveTransactionReview/index.js (hunk 5 above)
Comment (blackdevelopa) [reply app/components/UI/ApproveTransactionReview/index.js]: https://github.com/MetaMask/metamask-mobile/pull/5587/commits/850a93d1bae3298b38832c8e1446844ee7cbd8bc

File: app/components/UI/ApproveTransactionReview/index.js (hunk 4 above)
Comment (blackdevelopa) [reply app/components/UI/ApproveTransactionReview/index.js:74]: I think it's an ongoing conversation that we can adopt. The PS team may be enforcing this.

File: app/core/TransactionTypes.js [hunk 7]
```diff
@@ -9,6 +9,7 @@ export default {
     ETH: 'ETH',
     ERC20: 'ERC20',
     ERC721: 'ERC721',
+    ERC1155: 'ERC1155',
```
Comment (gantunesr) [inline app/core/TransactionTypes.js:12]: Is it possible to send ERC1155? Last time I checked this was not possible on mobile

File: app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js [hunk 8]
```diff
@@ -154,9 +163,16 @@ export default class TransactionReviewDetailsCard extends Component {
...
+            {tokenType === ERC20 && (
+              <Text style={styles.transactionDetailsTextRight}>
+                {formatNumber(allowance)} {tokenSymbol}
+              </Text>
+            )}
+            {tokenType === ERC721 && (
+              <Text style={styles.transactionDetailsTextRight}>
+                {tokenId} {tokenSymbol}
+              </Text>
+            )}
```
Comment (gantunesr) [inline app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js:175]: nit

```suggestion
            {tokenType === ERC20 ? (
              <Text style={styles.transactionDetailsTextRight}>
                {formatNumber(allowance)} {tokenSymbol}
              </Text>
            ) : null}
            {tokenType === ERC721 ? (
              <Text style={styles.transactionDetailsTextRight}>
                {tokenId} {tokenSymbol}
              </Text>
            ) : null}
```

**Review by gantunesr — COMMENTED** [review]
Some initial comments

**Comment by jpuri** [comment]
Changes look good to me, I would prefer to have a review from mobile team also before we merge bug PR like this.

File: app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js [hunk 9]
```diff
@@ -154,9 +163,16 @@ export default class TransactionReviewDetailsCard extends Component {
...
+            {tokenType === ERC20 ? (
+              <Text style={styles.transactionDetailsTextRight}>
+                {formatNumber(allowance)} {tokenSymbol}
+              </Text>
+            ) : null}
+            {tokenType === ERC721 ? (
+              <Text style={styles.transactionDetailsTextRight}>
+                {tokenId} {tokenSymbol}
+              </Text>
+            ) : null}
```
Comment (jpuri) [inline app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js:175]: Nit:
```
      <Text style={styles.transactionDetailsTextRight}>
        {tokenType === ERC20  ? formatNumber(allowance) : tokenId} {tokenSymbol}
      </Text>
```

**Comment by bschorchit** [comment]
> Changes look good to me, I would prefer to have a review from mobile team also before we merge bug PR like this.

@gantunesr could you give this a re-review once you have a chance? Thank you <3

File: app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js (hunk 9 above)
Comment (blackdevelopa) [reply app/components/UI/TransactionReview/TransactionReviewDetailsCard/index.js]: Good catch. Thank you. https://github.com/MetaMask/metamask-mobile/pull/5587/commits/bff82ce3b6a8669fc0f3877187c6796b8fc4c5ca

File: app/util/address/index.js [hunk 10]
```diff
@@ -437,3 +441,25 @@ export const stripHexPrefix = (str) => {
...
+      symbol,
+      standard,
+    };
+  }
+  return {
+    symbol,
+    decimals,
+    standard,
+  };
+};
```
Comment (gantunesr) [inline app/util/address/index.js:465]: Can you add some unit test for this method?

File: app/core/TransactionTypes.js (hunk 7 above)
Comment (gantunesr) [reply app/core/TransactionTypes.js:12]: Ah ok, that makes sense, I was thinking about the Send Flow.

File: app/components/Views/ApproveView/Approve/index.js [hunk 11]
```diff
@@ -40,23 +40,48 @@ import {
...
+      right: 0,
+      padding: 10,
+      color: colors.icon.default,
+    },
+    headerText: {
+      color: colors.text.default,
+      textAlign: 'center',
+      fontSize: 15,
+    },
+  });
```
Comment (gantunesr) [inline app/components/Views/ApproveView/Approve/index.js:84]: Move this to a styles file

**Review by gantunesr — APPROVED** [review]
LGTM. Left two non-blocker comments, you can do those changes in this PR or another one.

File: app/components/Views/ApproveView/Approve/index.js (hunk 11 above)
Comment (blackdevelopa) [reply app/components/Views/ApproveView/Approve/index.js]: https://github.com/MetaMask/metamask-mobile/pull/5587/commits/be6990bea85cf4c478f8a3ffabeb02d94cc5886d

**Comment by seaona** [comment]
I have some concerns about the ERC1155 feature. The ERC1155 standard has the `SetApprovalForAll` method available but notice the `approve` method for ERC721 does not apply here. See [here](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#ERC1155).

So I believe we should not do anything related to approving ERC1155 screen until we have the `SetApprovalForAll` screen in place. Does this make sense @blackdevelopa @bschorchit ?

**Comment by bschorchit** [comment]
> After reviewing this again, I have some concerns about the ERC1155 fix. In the ERC1155 standard there is no `approve` method. Instead we only have the `SetApprovalForAll` method available. See [here](https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#ERC1155).
> 
> So I believe we should not do anything related to approving ERC1155 screen until we have the `SetApprovalForAll` screen in place. Does this make sense @blackdevelopa @bschorchit ?

Interesting that it's not part of the standard - I have come across with some erc1155 that do implement approve method in the past. Currently for those cases on extension, we display the same screen we have for erc721 approve. I believe we could keep this same behavior for mobile, wdyt @seaona ?

**Comment by bschorchit** [comment]
> On the tx details, I believe we could remark this as a token ID? Something like using `#3` so it's not confused by an Amount? Or like we have on Extension, `Approved Asset: AssetName (# id)` (see right pic)
> 
> @bschorchit @blackdevelopa what do you think?
> 

I agree with you, @seaona! Could we make this update, @blackdevelopa ?

**Comment by seaona** [comment]
New changes look good! Great job @blackdevelopa !!
Thank you for your inputs @bschorchit !

https://user-images.githubusercontent.com/54408225/219042059-65883256-536a-4db6-8fbf-297e5add4174.mp4

**Pull request merged**